Add documentaion for advanced configuration of LDAP

Shubhamr837 opened the pull request:
Whats included in this pr ?

Documentation for how to configure the defaults of OpenLDAP .
Documentation for how to use another LDAP UI to access the OpenLDAP .
Documentation for how to change Sync-Endpoint to use/access a different LDAP than the OpenLDAP in the Docker swarm .
Documentation for using LDAP utils .

Should fix https://github.com/odk-x/tool-suite-X/issues/145

linl33 commented on odkx-src/sync-endpoint.rst:
```diff
@@ -132,3 +143,129 @@ Assigning users to groups
+       LDAP_ADMIN_PASSWORD=admin                  // default password for admin account 
+
+       # phpldapadmin
+       PHPLDAPADMIN_LDAP_HOSTS=ldap-service   // This is for the phpLDAPadmin . In docker swarm this is the name of the service running LDAP . This can be eddited
```
~~docker swarm~~ Docker Swarm
~~name~~ hostname

linl33 commented on odkx-src/sync-endpoint.rst:
```diff
@@ -132,3 +143,129 @@ Assigning users to groups
+
+Editing the defaults of LDAP directory
+"""""""""""""""""""""""""""""""""""""""""""""
+    Modify the ldap.env file to configure the environment variables . The default settings are as follows 
```
Could you also add a link to this file? So it's easier to find.

linl33 commented on odkx-src/sync-endpoint.rst:
```diff
@@ -132,3 +143,129 @@ Assigning users to groups
+   
+.. _sync-endpoint-custom-ldap:
+
+Using a Different LDAP directory
```
~~directory~~ Directory

linl33 replied:
The steps in the section assume the new LDAP server is also a Docker container, which seems to imply Sync Endpoint could only interact with an LDAP server hosted as a container. Please add a sentence or two to explain that the LDAP server could be hosted anywhere as long as the Sync Endpoint instance could communicate with it.

linl33 commented on odkx-src/sync-endpoint.rst:
```diff
@@ -132,3 +143,129 @@ Assigning users to groups
+       
+    4. Replace the ldap-service image from docker-compose.yml with odk/[LDAP_DIRECTORY_NAME].
+    
+    5. Modify config/sync-endpoint/security.properties to fill in the Settings for LDAP server.
```
Please add a link to this file.

linl33 commented on odkx-src/sync-endpoint.rst:
```diff
@@ -132,3 +143,129 @@ Assigning users to groups
+    4. Replace the ldap-service image from docker-compose.yml with odk/[LDAP_DIRECTORY_NAME].
+    
+    5. Modify config/sync-endpoint/security.properties to fill in the Settings for LDAP server.
+       Set security.server.ldapUrl in security.properties to the new server url . The name of the service in swarm would be same ( ldap-service ) . So just change 
```
This is not accurate. The preferences configured in `security.properties` (these lines: https://github.com/odk-x/sync-endpoint-default-setup/blob/0d97f322b6f3c48f86cdb47c9c1c7acf3f7bfb93/config/sync-endpoint/security.properties#L38-L51) all need to match the target LDAP server.

linl33 commented on odkx-src/sync-endpoint.rst:
```diff
@@ -132,3 +143,129 @@ Assigning users to groups
+
+Managing Identity through DHIS2
+"""""""""""""""""""""""""""""""""
+    Modify config/sync-endpoint/security.properties to fill in the Settings for DHIS2 Authentication section
```
The formatting of this paragraph looks a bit awkward. Was it supposed to be an ordered list?

linl33 replied:
Please add a link to this file.

linl33 commented on odkx-src/sync-endpoint.rst:
```diff
@@ -68,7 +79,7 @@ If you'd prefer to use the :program:`OpenLDAP` command line utilities, they're i
 
   .. code-block:: console
-
+ 
```
Please remove this extra space.

linl33 commented on odkx-src/sync-endpoint.rst:
```diff
@@ -132,3 +143,129 @@ Assigning users to groups
+    
+    If you want to use a UI outside the docker swarm in your local machine Modify the docker-compose.yml file in sync-endpoint-default-setup directory . Add ports 
+    mapping to the ldap service to expose the port 389 of ldap service to a port in your local host . If you wish to access 
+    the ldap protocol over TLS/SSL expose the port 636. Connect the UI application to this port on localhost.
```
Could you add a warning here that the LDAP service will not only be accessible from localhost but will also be exposed over the Docker ingress overlay network (which is exposed to the Internet in most cases).

linl33 commented on odkx-src/sync-endpoint.rst:
```diff
@@ -132,3 +143,129 @@ Assigning users to groups
+       LDAP_DOMAIN=example.org                    // domain of your organisation
+       LDAP_READONLY_USER=true                    // enable the read only user
+       LDAP_READONLY_USER_PASSWORD=readonly       // password for read only user
+       LDAP_ADMIN_PASSWORD=admin                  // default password for admin account 
```
Could you add a note here that says if these options are modified, the corresponding options in Sync Endpoint's `security.properties` need to be modified to match.

linl33 reviewed (changes requested):
Thanks for documenting this. The content is mostly good, please see the inline comments.
A few general comments:
 - In some of the text, there's an extra space immediately before the punctuation. For example,  
    > Set security.server.ldapUrl in security.properties to the new server url .  

    Please fix them.
 - Please try to use markup for commands, program names, file path, and code. For example, use ``:program:`OpenLDAP` `` instead of `OpenLDAP`. Use ``:file:`ldap.env` `` instead of `ldap.env`.
 - Please capitalize the proper nouns. e.g. Docker
 - Please avoid using both **_folder_** and **_directory_** in the same sentence. It is generally fine to use these 2 terms interchangeably but please stick to 1 in the same sentence.

Shubhamr837 commented:
@linl33 Thanks for your suggestions . I have tried to correct all the errors . Please review and suggest if further any change is required .

linl33 commented on odkx-src/sync-endpoint.rst:
```diff
@@ -132,3 +143,153 @@ Assigning users to groups
+
+.. _sync-endpoint-custom-ldap:
+
+Using a Different LDAP Directory
```
I just realized a similar section already exists at https://docs.odk-x.org/sync-endpoint-manual-setup/#custom-ldap-directory. Could you try to merge your content with the existing section?

linl33 commented on odkx-src/sync-endpoint.rst:
```diff
@@ -132,3 +143,153 @@ Assigning users to groups
+.. note::
+
+  The LDAP Directory here is configured to run inside the Docker Swarm. If you are running the LDAP Directory outside the Docker Swarm and it is accessible 
+  for the containers inside the Docker Swarm , you can directly follow step 5 to configure it .
```
> Swarm , you
> configure it .

Many of your sentences has extra spaces like these. I'm not sure if it's your editor adding them? Could you try to remove the extra spaces?

linl33 reviewed (changes requested):
Thanks for making the changes. Just a few more things.

Shubhamr837 commented:
Thanks for the suggestions @linl33 . I have removed those extra spaces and added the custom LDAP documentation to the sync-endpoint-manual-setup file . Please review and suggest if any change is required .

linl33 commented on odkx-src/sync-endpoint-manual-setup.rst:
```diff
@@ -151,14 +153,43 @@ Custom LDAP directory
 
-    b. Uncomment the relevant lines in the *configs* section in :file:`docker-compose.yml` and the *configs* section under the *sync* section in :file:`docker-compose.yml`.
+    b. Uncomment the relevant lines in the *configs* section in :file:`docker-compose.yml` and the *configs* section under the *sync* section in :file:`docker-
+       compose.yml`.
```
Extra space here.

linl33 reviewed (changes requested):
There's just 1 more small issue. Thanks for documenting this.

Shubhamr837 commented:
@linl33 Sorry for this error. I have corrected it. Please review. 
Thanks
